internal/agent/agent_test.go

```diff
@@ -113,6 +113,33 @@ var _ = Describe("Device Agent behavior", func() {
 			})
 		})
 
+		When("device/enrollmentrequests have been deleted", func() {
+			It("should ask for new enrollment", func() {
+				dev := enrollAndWaitForDevice(h, DefaultApproval())
+
+				// wait for the agent to retrieve the agent certificate from the EnrollmentRequest
+				Eventually(h.AgentDownloadedCertificate, TIMEOUT, POLLING).Should(BeTrue())
+
+				h.StopAgent()
+				_, err := h.Client.DeleteDevice(h.Context, *dev.Metadata.Name)
+				Expect(err).ToNot(HaveOccurred())
+
+				err = os.Remove(filepath.Join(h.TestDirPath, "agent.key"))
+				Expect(err).ToNot(HaveOccurred())
```

Comment:
Why would we remove the agent key from disk[1].  Isn't this what generates the CSR? If you delete this do you also change the fingerprint on reboot? 

failing test case

- restart the agent
- ensure fingerprint is not changed

Then in general I feel the agent should not react to unclear signals. In this case we are proposing replacing client side certs because of a possible human error server side. (delete csr) For this to work safely there should be very clear signal in the spec of intent. enrolled: false is an example I don't love it but its clear.

All changes to state during runtime should be explicit and deliberate/auditable. 

[1] https://github.com/flightctl/flightctl/blob/main/internal/agent/agent.go#L82

Reply:
Right, it is a weird situation, and now that I think about it... as soon as we delete the key, it doesn't,t matter if we delete the device/enrollment request or not.  

The agent, at this point, just feels it's not enrolled unless we persist some additional information to disk.

I started to think that we should ideally add another identity fingerprint related to the hardware (processor ID or anything that univocally tracks a device. so if an admin needs it can figure out that the same hardware was previously connected to a different device/enrollment request)

Reply:
Keep in mind we'll eventually use an agent.key generated and sealed in the TPM instead of the agent.key file on disk (though I may want to use that as fallback in case of TPM-less systems).

Then, consider that we *want* that an admin doesn't care about a real hardware identifier. If the agent.key ever needs to be deleted, e.g. as part of decommissioning, then,  yes, the device needs to be reenrolled with the new identity. Whether it's the same hardware shouldn't matter (but we _could_ probably determine that from the TPM chip if necessary).

Reply:
But well, I can envision cases where the admin cares if it is the same hardware or not. For example, if you want to keep track of specific hardware for any reason (block hardware that could have been compromised from re-joining in any way..., hardware that has anything very special.

Reply:
Thanks for having an eye @fzdarsky !